File: modules/kotlin-api/src/main/kotlin/org/llm4s/kotlin/Llm4s.kt
package org.llm4s.kotlin

import org.llm4s.java.JAgent
import org.llm4s.java.JLlmClient
import org.llm4s.java.LlmResult
import org.llm4s.java.Llm4s as JLlm4s

/**
 * Internal seam that allows tests to replace the calls to [JLlm4s] static
 * methods without requiring real LLM credentials in unit tests.
 */
internal interface ClientFactory {
    fun createDefault(): LlmResult<JLlmClient>
    fun createAgent(client: JLlmClient): JAgent
}

internal object DefaultClientFactory : ClientFactory {
    override fun createDefault(): LlmResult<JLlmClient> = JLlm4s.createDefaultClient()
    override fun createAgent(client: JLlmClient): JAgent = JLlm4s.createAgent(client)
}

/**
 * Entry-point factory for Kotlin callers.
 *
 * ```kotlin
 * val client = Llm4s.createDefaultClient()   // reads LLM_MODEL + API key env vars
 * val agent  = Llm4s.createAgent(client)
 * ```
 */
object Llm4s {
    internal var factory: ClientFactory = DefaultClientFactory

    /**
     * Creates an [LLMClientKt] from the default provider configured via
     * environment variables (e.g. `LLM_MODEL`, `OPENAI_API_KEY`).
     * Throws [LLMException] if the provider cannot be configured.
     */
    fun createDefaultClient(): LLMClientKt {
        val result = factory.createDefault()
        if (result.isSuccess) return LLMClientKt(result.get())
        val err = result.getError()
        throw LLMException(err.message ?: "Failed to create client", err)
    }

    /**
     * Wraps an [LLMClientKt] in an [AgentKt] ready for natural-language queries.
     */
    fun createAgent(client: LLMClientKt): AgentKt {
        val jAgent = factory.createAgent(client.underlying)
        return AgentKt(jAgent)
    }
}

File: modules/java-api/src/main/scala/org/llm4s/java/Llm4s.scala
package org.llm4s.java

import org.llm4s.agent.Agent
import org.llm4s.config.Llm4sConfig
import org.llm4s.llmconnect.LLMConnect
import org.llm4s.llmconnect.config.ProviderConfig

/**
 * Entry-point factory for Java callers.
 *
 * All methods return [[LlmResult]] so Java code never imports Scala `Either`
 * or deals with implicit/given parameters directly. The Scala-level
 * `ModelRegistryService` given is resolved internally and passed explicitly.
 *
 * === Quick-start (Java) ===
 * {{{
 * LlmResult<JLlmClient> r = Llm4s.createDefaultClient();
 * if (r.isSuccess()) {
 *     JLlmClient client = r.get();
 *     client.complete("Hello").ifSuccess(System.out::println);
 * }
 * }}}
 */
object Llm4s {

  /**
   * Creates a [[JLlmClient]] from the default provider configured via
   * environment variables (e.g. `LLM_MODEL`, `OPENAI_API_KEY`).
   */
  def createDefaultClient(): LlmResult[JLlmClient] = {
    val result = for {
      registry <- Llm4sConfig.modelRegistryService()
      config   <- Llm4sConfig.defaultProvider()
      client   <- LLMConnect.getClient(config)(using registry)
    } yield new JLlmClient(client)
    LlmResult.from(result)
  }

  /**
   * Creates a [[JLlmClient]] from an explicit [[ProviderConfig]].  Useful
   * when the caller constructs the config programmatically rather than relying
   * on environment variables.
   */
  def createClient(config: ProviderConfig): LlmResult[JLlmClient] = {
    val result = for {
      registry <- Llm4sConfig.modelRegistryService()
      client   <- LLMConnect.getClient(config)(using registry)
    } yield new JLlmClient(client)
    LlmResult.from(result)
  }

  /**
   * Wraps a [[JLlmClient]] in a [[JAgent]] ready to accept natural-language
   * queries and call tools.
   */
  def createAgent(client: JLlmClient): JAgent =
    new JAgent(new Agent(client.underlying))
}
